src/wallet/wallet.cpp

```diff
@@ -4025,9 +4025,6 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
             pindexRescan = FindForkInGlobalIndex(chainActive, locator);
     }
 
-    walletInstance->m_last_block_processed = chainActive.Tip();
-    RegisterValidationInterface(walletInstance);
```

Comment:
I don't think you want to move the m_last_block_processed line above, because rescan code below will access it.

I'm also not sure it's a great idea to be moving the `RegisterValidationInterface` call. It may work for now, but create a race condition if we allow loading wallets after node startup (see #10740).

I think I'd suggest a more limited change than what's implemented here:

* Creating the wallet above with:

```c++
auto wallet_deleter = MakeUnique<CWallet>(...);
CWallet* walletInstance = wallet_deleter.get();
```

* And changing this line to:

```c++
RegisterValidationInterface(wallet_deleter.release());
```

Reply:
I don't see where `m_last_block_processed` is accessed in the rescan code. It seems to me it's just used to implement `BlockUntilSyncedToCurrentChain`.

The RegisterValidationInterface call is tricky. I agree moving it down only works if the wallets are all loaded *before* the Connman starts, as is the case right now, but it would break if wallets are loaded/unloaded via RPC. Once way to handle it could be to call `UnregisterValidationInterface` in the `CWallet` destructor (even without calling `RegisterValidationInterface` in the constructor). In that case, we could leave the register call where it is. What do you think of that?

Reply:
https://github.com/bitcoin/bitcoin/pull/12647#discussion_r174875670

> I don't see where m_last_block_processed is accessed in the rescan code. It seems to me it's just used to implement BlockUntilSyncedToCurrentChain.

You're right. I misremembered and it should be fine to move this, if that's desirable.

>  Once way to handle it could be to call UnregisterValidationInterface in the CWallet destructor (even without calling RegisterValidationInterface in the constructor). In that case, we could leave the register call where it is. What do you think of that?

Seems good. Would fix the immediate problem and also help with dynamic loading.